# STA 130 Week 01 Homework

### 1. Picking dataset

In [45]:
# using example data set
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()          #chatbot states id column has 1 value missing, song has 11 values missing

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

### 2. Prompting chatbot and requesting num of columns/rows

In [46]:
df.shape            #(num rows, num columns)

(391, 11)

In [47]:
len(df.columns)     #num of columns

11

In [48]:
len(df)             #num of rows

391

In [49]:
df.columns          #names of columns

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

In [50]:
columns = df.columns#same result as above (uploaded DF to chatbot so it prompted me with code specific to a local file)
print(columns)

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


In [51]:
df.info()          #quick overview of data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


#### Observations / Variables

ChatGPT understood Observation to represent one data entry, with each row corresponding to a single villager in Animal Crossing with info about name, species, birthdays, other info, etc. Number of rows = number of individual villagers described.

Variables were columns which represents info about our observation/row/villager including name, species, gender, birthday etc.

Thus,
- observations = rows
- variables = columns  
in the context of our pandas dataset.

### 3. Asking for summary of columns & suggested code to do so

In [52]:
df.describe()          #summary stats for numerical columns

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [53]:
#summary stats for categorical (non-num columns)
df.describe(include='object')

id     name gender species birthday personality          song  \
count       390      391    391     391      391         391           380   
unique      390      391      2      35      361           8            92   
top     admiral  Admiral   male     cat     1-27        lazy  K.K. Country   
freq          1        1    204      23        2          60            10   

         phrase           full_id  \
count       391               391   
unique      388               391   
top     wee one  villager-admiral   
freq          2                 1   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1

In [54]:
#shows how often each unique value appears in a particular column
for col in df.select_dtypes(include='object').columns:
    print(f"Value counts for {col}:")
    print(df[col].value_counts())
    print("\n")

Value counts for id:
id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64


Value counts for name:
name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64


Value counts for gender:
gender
male      204
female    187
Name: count, dtype: int64


Value counts for species:
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger     

### 4. Switching data sets & asking for discrepancy clarification

In [55]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(url)

In [56]:
#891 rows (observations), 15 columns(variables)
titanic_df.shape

(891, 15)

In [57]:
#gives summary stats (count, mean, std dev) but only to num columns unless include='all' is specified 
titanic_df.describe(include='all')
#all 15 columns are displayed here
#num columns being those of type float or int

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

In [58]:
#when ran without (include='all'), categorial non-numerical columns are omitted as shown
titanic_df.describe()
#only 6 of the 15 columns are displayed here

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

ChatGPT also mentions the function df.describe() only calculates stats for non-missing values in each column. Thus, for rows with missing vallues, the number of rows in df.describe() is less than df.shape().

In [59]:
#running df.tail() to see proof of 891 rows (starts at 0 so 1+890=891)
titanic_df.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

I asked ChatGPT why missing values in data sets are referred to as **NaN** and I learned it stands for **"Not a Number"**, based on the IEE 754 standard for floating-point arithmetic.

In [60]:
#.dropna() drops any row with at least 1 missing value
cleaned_df = titanic_df.dropna()
cleaned_df.describe()

survived      pclass         age       sibsp       parch        fare
count  182.000000  182.000000  182.000000  182.000000  182.000000  182.000000
mean     0.675824    1.192308   35.623187    0.467033    0.478022   78.919735
std      0.469357    0.516411   15.671615    0.645007    0.755869   76.490774
min      0.000000    1.000000    0.920000    0.000000    0.000000    0.000000
25%      0.000000    1.000000   24.000000    0.000000    0.000000   29.700000
50%      1.000000    1.000000   36.000000    0.000000    0.000000   57.000000
75%      1.000000    1.000000   47.750000    1.000000    1.000000   90.000000
max      1.000000    3.000000   80.000000    3.000000    4.000000  512.329200

### 5. Attributes vs Methods / Function Calls

In [61]:
titanic_df.describe          #attribute (no parentheses)

<bound method NDFrame.describe of      survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult

In [62]:
titanic_df.describe()        #method/function call (with parentheses)

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- Attribute  
An attribute simply holds or returns a value, it does not perform any action. Thus, attributes are characteristics or properties of an object. In the context of a pandas data frame, df.describe simply refrences the describe method, but doesn't execute it. As such, the info df.describe() contains when refrencing our data set is displayed as raw, unformatted text.

- Methods  
Methods perform an action when called upon, which can be done by adding parentheses() after a function name which executes the function. When df.describe() is ran, the python function is executed and the information from df.describe() when refrencing our data set is formatted and displayed neatly in a table.

### Summary of pre-lecture questions

https://chatgpt.com/share/a3945623-ec44-4ccd-8573-0c4a0bc1b56e

Chatbot Session Summary
Topics Covered:
1. Checking Data Set Columns and Size
To find the columns of a pandas DataFrame, use the columns attribute to view all column names.
To check the size of the dataset, use the shape attribute, which returns a tuple with the number of rows and columns.
2. Attributes vs Methods
Attributes are properties of objects (e.g., df.shape) and do not require parentheses.
Methods are functions associated with objects (e.g., df.describe()), and require parentheses for execution.
3. Discrepancy Between df.shape and df.describe()
df.shape provides the total number of rows and columns in the dataset, including all data types.
df.describe() provides summary statistics only for numeric columns by default, such as the count, mean, and standard deviation. This can lead to a discrepancy where df.shape counts more columns than df.describe() because non-numeric columns (like strings or categories) are excluded from the summary.
To include all columns (including categorical ones) in describe(), use the argument include='all'.
4. Handling Missing Values (NaN)
Missing values in datasets are often represented as NaN (Not a Number). This allows for consistent numerical operations despite missing data.
You can use functions like dropna() to remove rows with missing values or describe() to summarize non-missing data.
5. Differences Between Mathematical Functions and Programming Functions
Mathematical Functions: Map inputs to outputs with a clear, deterministic relationship (e.g., f(x)=x^2).
Programming Functions: Blocks of reusable code that perform tasks, which can include side effects like modifying variables or printing output.
Aspect	Mathematical Function	Programming Function
Purpose	Maps input to output through a rule	Executes a task with or without returning output
Determinism	Always returns the same result for the same input	May return different results or have side effects
Return Values	Always returns one output	May return one or more values, or none
For more detailed explanations, you can check the data sets and examples discussed during our session.


### 6. Definitions of rows provided by describe() method

In [63]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(url)

titanic_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

##### Definitions + examples of df.describe() rows according to ChatGPT:
- count: number of non-null (non-NaN) entries in the column
    - ex. if a column has 100 rows but 2 rows contain missing values (NaN), *count* = 98.
    
- mean: average of the values in the column, calculated as the sum of all values divided by the *count*.
    - ex. if the values in a column are [2, 3, 5], the mean would be (2+3+5)/3 = 3.33
    
- std (standard deviation): measure of the dipersion/spread of the values from the mean. it indicates how much the values typically deviate from the mean.
    - ex. low std: values are closely packed around the mean while
    - ex. high std: values are spread out further from mean
    
- min: smallest (minimum) value in the column
    - ex. if values in a column are[2, 5, 3], the *min* would be 2.
    
- 25% (1st quartile / Q1): refers to the 25th percentile of data, the value below which 25% of the data fall.
    - ex. if 25% of the values in a column <=2, the 25% value would be 2.

- 50% (2nd quartile / Q2): refers to the 50th percentile of data, half of the values are below this point, while half are above. also known as the **median**.
    - ex. if values in a column are [1, 3, 5] the median would be 3.
    
- 75% (3rd quartile / Q3): refers to the 75th percentile of the data, the value below which 75% of the dat fall.
    - ex. if 75% of the values in a column >=5, the 75% value would be 5.
    
- max: largest (maximum) value in the column
    - ex. if values in a column are [2, 3, 5], the *max* would be 5.


#### What does df.describe() do with missing values?

- When df.describe() runs in pandas, it automatically **ignores missing values (NaNs)** when calculating summary statistics.

- Count reflects only number of non-NaN values in each column
- All other summary stats are calculated based on available values (NaN's are ignored)


### 7. Using df.dropna() or del df['col'] to maximize efficiency

In [64]:
titanic_df.dropna()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
871  woman       False    D  Southampton   yes  False  
872    man        True    B  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[182 rows x 15 columns]

- df.dropna()
    - Removes rows or columns cointaing missing values, allowing one to retain as much valid data as poppsible in the Data Frame.
    - Can selectively remove only rows where a specific column (or set of columns) has missing values using options like *subset* or *how*.


In [65]:
#df.dropna()  # Drops rows with any missing values
titanic_df.dropna(subset=['pclass', 'age'])  # Drops rows only if 'col1' or 'col2' has missing values

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
885  woman       False  NaN   Queenstown    no  False  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[714 rows x 15 columns]

In [66]:
titanic_df.shape    #confirming titanic_df still contains full original data

(891, 15)

##### df.dropna() over del df['col']
- As shown above when running attribute df.shape, the titanic data frame has a total of 891 rows. However, when df.dropna() is ran, only 182 rows are displayed as rows as all rows with missing data is dropped.
- An advantage of df.dropna(), is that we can **selectively remove rows** as shown above with (subset=['pclass', 'age'] as it drops rows only if 'col1' (pclass) and 'col2' (age) have missing values. Thus, 714 rows are now displayed when subset is ran.

##### del df['col'] over df.dropna()

- One might choose del df['col'] over df.dropna() to remove an entire column regardless of NaN's.
- Entire columns can be removed to **maximize efficiency** and **simplify data preprocessing** if not needed (e.g. identifier, metadata, any other irrelevant feature) or if a column contains significant number of missing values to the point where available data is sparse not usable.
- dropna() may remove too many rows in the case of a column having many missing values. In this case, simply removing a problematic column with **del** may be better to **preserve rows of valuable data** in other columns.

##### Importance of applying del df['col'] before using df.dropna()

- Primary factor is  of **reduction of data loss**
    - Which helps maintain data integrity & simplifying data cleaning.
    - If a particular column has high number of missing values, applying dropna() first may cause an unnecessary drop in rows resulting in signicant loss of useful data from other columns
    - Using del df['col'] first ensures avoiding dropping rows that might contain valuable info in other columns.
    - ex. If 'col' has ~80% missing values, using df.dropna() *first* would remove ~80% of dataset. easily avoided by simply deleting the problematic column before dropping rows.

##### Removing all missing data from titanic data set

In [67]:
titanic_df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [68]:
# checking for missing values (age - 177 / deck - 688 / embark_town - 2)
titanic_df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [70]:
# removing 'deck' due to excessive missing data
del titanic_df['deck']

In [72]:
titanic_df #checking to see if deck was removed

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

In [87]:
#dropping rows with remaining missing data using df.dropna() and assigning new variable name to cleaned data set
titanic_df_cleaned = titanic_df.dropna()

In [85]:
#checking to see if remaining missing data was removed
titanic_df_cleaned.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

- The result of .isna().sum of the cleaned data set differs greatly from the result of .isna().sum of the original data set, indicating that the cleaned data set does not inlude any missing data.

#### Summary of questions 6 and 7

https://chatgpt.com/share/cc9c8318-0a9b-4b9d-aafb-5653655012d6

Here's a concise summary of our conversation:

df.describe() Statistics:

We discussed the terms provided by the df.describe() method: count, mean, std, min, 25%, 50%, 75%, and max. These statistics give insights into the distribution, central tendency, and spread of data while excluding missing values.


Handling Missing Data:


We explored the difference between using df.dropna() and del df['col']. df.dropna() removes rows or columns with missing data, whereas del removes an entire column.
Preference for df.dropna(): It's used when you want to selectively remove missing data from rows or specific columns without discarding entire columns that may still contain useful data.
Preference for del df['col']: This is preferred when a column is irrelevant or contains too many missing values, making it more efficient to remove the entire column rather than losing rows from the dataset.


Efficiency of Combining del and df.dropna():


Using del df['col'] first helps prevent the removal of too many rows when using df.dropna(), especially if a column has excessive missing values.
By deleting columns with high proportions of missing data, you ensure that subsequent calls to df.dropna() preserve more rows, making the process more efficient.


Efficient Removal of Missing Data in the Titanic Dataset:


We discussed a strategy for handling missing data in the Titanic dataset by first inspecting missing values, deleting columns with excessive missing data (like 'deck'), and then applying df.dropna() to remove rows with missing data in the remaining columns.
This approach ensures that the dataset is cleaned efficiently without losing too much valuable information.

### 8. Give brief explanations for any requested answers to the questions below

In [69]:
titanic_df.groupby("who")["age"].describe()

count       mean        std    min   25%   50%   75%   max
who                                                              
child   83.0   6.369518   4.729063   0.42   2.0   5.0   9.5  15.0
man    413.0  33.173123  12.906908  16.00  23.0  30.0  40.0  80.0
woman  218.0  32.000000  11.495441  16.00  23.0  30.0  39.0  63.0

1. titanic_df.groupby("who") groups the data set by the unique values in column 1, with each unique value defining a group. In this case, the groups are: child, man, woman. Then the .describe() method generates summary statistics for values of ["age"]. For example, the minimum age to be considered an adult (non-child) is 16.

In [90]:
titanic_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

2. titanic_df.describe() generates different values for count than titanic_df.groupby("who")["age"].describe() as df.describe() provides statistics for the entire data set. In contrast, the command above limits our statistical summary to the groups specified as values under the who column. Thus, count is only considered for values under

3. Intentionally Introducing Errors

In [3]:
pd

NameError: name 'pd' is not defined

a) ChatGPT required one less click and no scrolling to find an answer. It immeditely assumed pd had yet to be imported. When googling I was faced with numerous Stack Overflow posts which I'd have to filter through to find my answer. Therefore, ChatGPT was much quicker than google.

In [11]:
import pandas as pd
pd.read_csv("titanics.csv")     

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

b) ChatGPT once again immeditely assumed the issue correctly and prompted me to ensure my dataset was named correctly, and that I was pointing to the correct file path. Google required filtering through various Stack Overflow posts, therefore ChatGPT wins again.

In [15]:
df = pd.read_csv("titanic.csv") #pointing to correct file path
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [16]:
DF.groupby("class")["sex"].describe()

NameError: name 'DF' is not defined

c) ChatGPT once again correctly assumed my typo, stating the correct variable name should be df (lowercase). Google requires filtering through numerous Stack Overflow posts, which vary in quality. ChatGPT wins again.

In [18]:
df.groupby("class")["sex"].describe() #using correct variable name

count unique   top freq
class                         
First    216      2  male  122
Second   184      2  male  108
Third    491      2  male  347

In [21]:
pd.read_csv("titanic.csv"

SyntaxError: incomplete input (398260284.py, line 1)

d) Without specifying the code, ChatGPT correctly guessed the issue possibly being an open parenthesis or quotation mark without a closing counterpart. Therefore, ChatGPT is easier as googling requires filtering through various Stack Overflow (and/or other forums) for answers of various quality.



In [25]:
pd.read_csv("titanic.csv") #inputting code with closed parenthesis

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [26]:
df.group_by("class")["fare"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [30]:
df.groupby("class")["fare"].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

e) ChatGPT continues it's winning streak by correctly troubleshooting my error and altering me that the correct method names are 'groupby' and 'describe()'. As mentioned above, googling requires a bit more effort to troubleshoot this error. Therefore, ChatGPT is faster.

In [31]:
df.groupby("class")["fare"].describe() #using correct method names

count       mean        std  min       25%      50%   75%       max
class                                                                      
First   216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
Second  184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
Third   491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

In [29]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [33]:
df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

f) ChatGPT again corrects my error with Usain Bolt like speed, indicating that the data set I provided uses lowercase column names, and stating that 'sex' and 'age' should work. Googling shows numerous links to Stack Overflow, Reddit, and other such forums which I must comb through to trouble shoot my issue. Thus, ChatGPT is faster once again.

In [35]:
df.groupby("sex")["age"].describe() #using correct column names

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [37]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

In [38]:
df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

g) ChatGPT alerts me that the errors occur as Pandas treats my column names without quotes as variables instead of their intended purpose as columns, in which they must be specified as strings through quotation marks. Google, once again lead me to various forum posts with code unrelated to my own data set that I am working with. Therefore, much like the other intentional errors, ChatGPT was much faster and efficient with troubleshooting.

### 9. Revewing wiki and interacting with ChatBot for clarification

Yes, I have familiarized myself with most of the content of the wiki-textbook and used ChatBots when I ran into something which needed clarification.

### Summary of question 8

https://chatgpt.com/share/920a4dde-b7a9-4005-b9db-c02002194a57
    
Summary of Titanic Dataset Analysis
Loading the Dataset:

The Titanic dataset was accessed via a URL: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv.
The dataset was loaded into a Pandas DataFrame using the pd.read_csv(url) method.


Common Errors Encountered:


NameError: name 'pd' is not defined: This error occurred because the Pandas library had not been imported. The solution was to add import pandas as pd at the top of the script.


FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv': This error was due to the incorrect file path. The solution was to use the URL directly to load the dataset.


AttributeError: 'DataFrame' object has no attribute 'group_by': The method group_by was incorrectly used instead of groupby. The correct method is groupby.


KeyError: 'Sex': The column name was case-sensitive. The correct column name in the dataset is sex.


AttributeError: 'SeriesGroupBy' object has no attribute 'describle': There was a typo in the method name. The correct method is describe().


KeyError: 'Column not found: Age': This was due to a case sensitivity issue. The correct column name is age.